### PR TITLE
2021 10 merging

### PR DESCRIPTION
**Pull request by mathrack**

### CODE REVIEW

**Comment by mathrack** [comment]
The changelog summarizing the modifications is [here](https://github.com/mathrack/Incompact3d-1/blob/2021_10_merging/Changelog). Reproduced here with a task list. We can check / uncheck the modifications and I will try to update the PR accordingly.

Major updates :
- [x] Check most MPI and ADIOS2 error codes
- [x] Separate IBM reconstruction and derive / filter subroutines. The subroutines der* and fil* no longer perform IBM reconstruction. This combined operation can be obtained using the subroutines ibm_der* and ibm_fil*
- [x] Thomas tri-diagonal solver is moved in a dedicated module
- [x] Remove most of the memory leaks
- [x] Add unit tests for all derive and filter subroutines (depends on pfunit)
- [x] Avoid IO with pre-defined unit number
- [x] Add boot & finalize subroutines for each case to open and close IO units
- [x] Add a differentially heated cavity example
- [x] Move subroutines from the file xcompact3d.f90 to x3d_tools.f90. The file xcompact3d.f90 contains only the program.
- [x] X3D can catch the signals SIGUSR1 and SIGUSR2

Minor updates :
- [x] indent.sh : force homogeneous if statements in the code
- [x] Makefile : explicit file dependency allow parallel make
- [x] decomp2d/decomp2d.f90 : use iso_fortran_env for single / double precision kind
- [x] decomp2d/decomp2d.f90 : refactor transpose operations to avoid the optional argument (some compilers allocate memory for the optional argument at each call otherwise)
- [x] decomp2d/decomp2d.f90 : new error subroutine provides the file and line to locate the error
- [x] decomp2d/io.f90 : add read_plane
- [x] The module src/ibm_param is merged inside the module src/ibm
- [x] src/BC-Channel.f90 : try to initialize the velocity with a binary file first
- [x] src/BC-Channel.f90 : monitor bulk quantities by default
- [x] src/BC-dbg-schemes.f90 : inhouse sin / cos / ... functions can be applied to arrays using the elemental keyword
- [x] src/case.f90/postprocessing : move the potentially large array T_tmp to the var module
- [x] src/case.f90/postprocess_case : the subroutine force was called with the wrong number of arguments
- [x] src/derive.f90 : major refactoring
- [x] src/filters.f90 : major refactoring
- [x] src/forces.f90 : Move the subroutine force inside the module
- [x] src/ibm.f90 : add subroutines apply_ibm* for IBM reconstruction
- [x] src/implicit.f90 : unused 1D arrays are deallocated at the beginning of the simulation
- [x] src/les_models.f90 : local 3D arrays wallflux*1 are moved inside the module var
- [x] src/les_models.f90 : sgs_scalar_nonconservative: use pre-allocated arrays phi2 and phi3
- [x] src/module_param.f90 : der* and fil* subroutines can no longer modify the input 3D array. This should allow OCC.
- [x] src/module_param.f90 : uvwt_*bound are now used for out of bound detection
- [x] src/navier.f90 : compute max(abs(div(u))) instead of max(div(u))
- [x] src/navier.f90 : div(u*) and div(u) in the listing are no longer changing when the 2D pencil configuration (or the number of CPUs) is modified
- [x] src/parameters.f90 : initstat set to ifirst when initstat < 0
- [x] src/parameters.f90 : print all the detected compile flags in the listing
- [x] src/parameters.f90 : current implicit time scheme is not compatible with IIBM=3
- [x] src/probes.f90 : flag_* are initialized in parameters.f90
- [x] src/statistics.f90 : add the variance of the pressure
- [x] src/statistics.f90 : fix a bug, nvisu was used instead of nstat
- [x] src/tools.f90 : add a function to compute the L1 / L2 / Linf errors
- [x] src/tools.f90 : min_max detection based on uvwt_*bound arrays
- [x] src/tools.f90 : when min/max is reached, all CPUs call MPI_ABORT
- [x] src/transeq.f90 : change intent(in) for some arrays (in case of IBM)
- [x] src/variables.f90 : do not allocate all the variables all the time (statistics or ABL arrays are not always needed)
- [x] examples/cavity : add input file and a python SPOD script
- [x] examples/channel : python script to plot the channel.dat values
- [x] examples/channel : Slurm example to send SIGUSR2
- [x] examples/TGV : input files for 2D TGV
- [x] tests/verif : unit tests with verification of the error level for each derive and filter subroutine.
- [x] tests/convergence : unit tests without verification of the error level for each derive and filter subroutine (mesh convergence)

TODO list for this PR :
- [x] Makefile : add `-ftree-vectorize` in the GNU section
- [x] decomp2d/decomp2d.f90 : error subroutines should write to both stdout and stderr
- [x] Move the new 2D TGV input files to a folder dedicated to verification of the code
- [ ] Full support of the unit tests in the cmake build

File: src/transeq.f90 [hunk 1]
```diff
@@ -764,10 +764,8 @@ subroutine scalar_transport_eq(dphi1, rho1, ux1, uy1, uz1, phi1, schmidt, is_eve
       tb1(:,:,:) = ux1(:,:,:) * tb1(:,:,:)
     endif
     if (skewsc) then
-      ! (1 - 0.5) * ux.dTdx + 0.5 * (d(ux.T)dx - T.duxdx)
+      ! (1 - 0.5) * ux.dTdx + 0.5 * d(ux.T)dx
```
Comment (slaizet) [inline src/transeq.f90]: should be (one-half) and half

File: src/transeq.f90 [hunk 2]
```diff
@@ -836,10 +834,8 @@ subroutine scalar_transport_eq(dphi1, rho1, ux1, uy1, uz1, phi1, schmidt, is_eve
        tb2(:,:,:) = uy2(:,:,:) * tc2(:,:,:)
     endif
     if (skewsc) then
-      ! (1-0.5) * uy.dTdy + 0.5 * (d(uy.T)dy - T.duydy)
+      ! (1-0.5) * uy.dTdy + 0.5 * d(uy.T)dy
```
Comment (slaizet) [inline src/transeq.f90]: should be (one-half) and half

File: src/transeq.f90 [hunk 3]
```diff
@@ -865,10 +861,8 @@ subroutine scalar_transport_eq(dphi1, rho1, ux1, uy1, uz1, phi1, schmidt, is_eve
       tb3(:,:,:) = uz3(:,:,:) * tb3(:,:,:)
     endif
     if (skewsc) then
-      ! (1-0.5) * uz.dTdz + 0.5 * (d(uz.T)dz - T.duzdz)
+      ! (1-0.5) * uz.dTdz + 0.5 * d(uz.T)dz
```
Comment (slaizet) [inline src/transeq.f90]: should be (one-half) and half

**Review by slaizet — CHANGES_REQUESTED** [review]

File: src/transeq.f90 (hunk 1 above)
Comment (mathrack) [reply src/transeq.f90]: This is fixed in the updated version of this PR

File: src/transeq.f90 (hunk 2 above)
Comment (mathrack) [reply src/transeq.f90]: This is fixed in the updated version of this PR

File: src/transeq.f90 (hunk 3 above)
Comment (mathrack) [reply src/transeq.f90]: This is fixed in the updated version of this PR

**Review by mathrack — COMMENTED** [review]
The modifications correspond to old commits. This part of the code no longer exists. The requested change can not be applied.

**Pull request closed without merging**